sklearn/neighbors/unsupervised.py

```diff
@@ -122,3 +125,149 @@ def __init__(self, n_neighbors=5, radius=1.0,
+          'yule']
+
+        See the documentation for scipy.spatial.distance for details on these
+        metrics.
+
+    p : integer, optional (default = 2)
+        Parameter for the Minkowski metric from
+        sklearn.metrics.pairwise.pairwise_distances. When p = 1, this is
+        equivalent to using manhattan_distance (l1), and euclidean_distance
+        (l2) for p = 2. For arbitrary p, minkowski_distance (l_p) is used.
+
+    metric_params : dict, optional (default = None)
+        Additional keyword arguments for the metric function.
+
+    n_jobs : int, optional (default = 1)
+        The number of parallel jobs to run for neighbors search.
+        If ``-1``, then the number of jobs is set to the number of CPU cores.
+        Affects only :meth:`kneighbors` and :meth:`kneighbors_graph` methods.
+
+    Examples
+    --------
+    >>> from sklearn.cluster import DBSCAN
+    >>> from sklearn.neighbors import NearestNeighborsTransformer
+    >>> from sklearn.pipeline import make_pipeline
+    >>> estimator = make_pipeline(
+    ...     NearestNeighborsTransformer(radius=42.0, mode='distances'),
+    ...     DBSCAN(min_samples=30, metric='precomputed'))
+    """
+    def __init__(self, mode='connectivity', include_self=None,
+                 n_neighbors=None, radius=None, algorithm='auto', leaf_size=30,
+                 metric='minkowski', p=2, metric_params=None, n_jobs=1):
+        super(NearestNeighborsTransformer, self).__init__(
+            n_neighbors=n_neighbors, radius=radius, algorithm=algorithm,
+            leaf_size=leaf_size, metric=metric, p=p,
+            metric_params=metric_params, n_jobs=n_jobs)
+        self.mode = mode
+        self.include_self = include_self
+
+    def transform(self, X):
+        """Computes the (weighted) graph of Neighbors for points in X
+
+        Parameters
+        ----------
+        X : array-like, shape = [n_samples_transform, n_features]
+            Sample data
+
+        Returns
+        -------
+        Xt : CSR sparse matrix, shape = [n_samples_fit, n_samples_transform]
+            Xt[i, j] is assigned the weight of edge that connects i to j.
+        """
+        check_is_fitted(self, '_fit_X')
+        check_array(X, accept_sparse='csr')
+        if self.radius is None and self.n_neighbors is None:
+            warnings.warn("Please specify either radius or n_neighbors. "
+                          "Default to n_neighbors = 5.", UserWarning)
+            n_neighbors = 5
+        else:
+            n_neighbors = self.n_neighbors
+
+        if self.radius is not None and self.n_neighbors is not None:
+            raise ValueError(
+                "Please do not specify both radius and n_neighbors.")
+
+        if self.include_self is None:
+            include_self = self.mode == 'connectivity'
+        else:
+            include_self = self.include_self
+
+        # If we don't include each sample as its own neighbors
+        if not include_self and X is self._fit_X:
```

Comment:
No, it's not sane to condition this on identity. Use X=None in fit_transform only.